File: 321-bisect-left-right.py

#given one array,likes [1,2,2,3,3,3,4,4,5], bisect_left(array,3) will return the index of 3

#The purpose of Bisect algorithm is to find a position in list where an element needs to be inserted to keep the list sorted.

#Python in its definition provides the bisect algorithms using the module “bisect” which allows to keep the list 
# in sorted order after insertion of each element. 
# This is essential as this reduces overhead time required to sort the list again and again 
# after insertion of each element.

"""
bisect_left(list, num, beg, end) :- This function returns the position in the sorted list, 
        where the number passed in argument can be placed so as to maintain the resultant list in sorted order. 
        If the element is already present in the list, the left most position where element has to be inserted is returned.
        This function takes 4 arguments, list which has to be worked with, 
        number to insert, starting position in list to consider, ending position which has to be considered.
"""
def bisec_left(arr,x):
    for i in range(len(arr)):
        if arr[i]>=x:
            return i
    return len(arr)



def bisec_right(arr,x):
    for i in range(len(arr)):
        if arr[i]>x:
            return i
    return len(arr)

def bisect_right(a,x,lo=0,hi=None):
    if lo<0:
        raise ValueError("Wrong Value of lo\n")
    if hi is None:
        hi=len(a)
    while lo<hi:
        mid=(lo+hi)//2
        if a[mid]>x:
            hi=mid
        else:
            lo=mid+1
    return lo

def bisect_left(a,x,lo=0,hi=None):
    if lo<0:
        raise ValueError("Wrong Value of lo.\n")
    if hi is None:
        hi=len(a)
    while lo<hi:
        mid=(lo+hi)//2
        if a[mid]>=x:
            hi=mid
        else:
            lo=mid+1
    return lo
    
arr=[1,2,2,3,3,3,4,4,5]
# print(bisec_right(arr, 0))
print(bisect_right(arr, 3,0,6))  #return the right-most pos 
print(bisect_left(arr, 3,0,1))   #return the left-most position

File: 321-bisect-left-right.cpp
//given one sorted array and new value, return the index of this new value in array
#include<iostream>
#include<vector>

using DT=std::vector<int>;

int bisect_left(const DT& a,int val)
{
    if(a.empty())
        return 0;
    int sz=a.size();
    for(int i=0;i<sz;++i)
    {
        if(a[i]>=val)
            return i;
    }
    return sz;
}

int bisect_right(const DT& a,int val)
{
    if(a.empty())
        return 0;

    int sz=a.size();
    for(int i=0;i<sz;++i)
    {
        if(a[i]>val)
            return i;
    }
    return sz;
}

int bisect_right(const DT& a,int val,int lo=0,int hi=0)
{
    if(lo<0)
        return -1;
    if(hi==0)
        return a.size();
    
    while(lo<hi)
    {
        int mid=lo+(hi-lo)/2;
        if(a[mid]>val)
            mid=hi;
        else
            lo=mid+1;
    }
    return lo;
}

int main()
{
    DT a{1,2,2,3,3,3,4,4,5};
    int idx=bisect_left(a,3);
    // int ridx=bisect_right(a,3);
    int rridx=bisect_right(a,3,6,a.size());
    std::cout<<idx<<'\n';
    // std::cout<<ridx<<'\n';
    std::cout<<rridx<<'\n';
    return 0;
}